pkg/operator/context.go

```diff
@@ -0,0 +1,114 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package operator
+
+import (
+	"context"
+	"runtime/debug"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/client-go/kubernetes"
+	clientgoscheme "k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/util/flowcontrol"
+	"k8s.io/utils/clock"
+	"knative.dev/pkg/configmap/informer"
+	knativeinjection "knative.dev/pkg/injection"
+	"knative.dev/pkg/logging"
+	"knative.dev/pkg/system"
+	controllerruntime "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+
+	"github.com/aws/karpenter/pkg/apis"
+	"github.com/aws/karpenter/pkg/config"
+	"github.com/aws/karpenter/pkg/events"
+	"github.com/aws/karpenter/pkg/utils/injection"
+	"github.com/aws/karpenter/pkg/utils/project"
+)
+
+const (
+	appName   = "karpenter"
+	component = "controller"
+)
+
+var (
+	scheme = runtime.NewScheme()
+)
+
+func init() {
+	utilruntime.Must(clientgoscheme.AddToScheme(scheme))
+	utilruntime.Must(apis.AddToScheme(scheme))
+}
+
+// Context is the root context for the operator and exposes shared components used by the entire process
+type Context struct {
```

Comment:
After some thought, I want to discuss this decision a bit more. Primarily because I am concerned with the ramifications of passing fields that would normally be passed through to controller constructors as part of the `context.Context` struct.

In general, the `Context` object is used for request-specific context and cancellation and not for instantiation/performing dependency injection. 

I'm curious for what your thought is in where else this `Context` might be passed; otherwise, I'm leaning more towards continuing calling this `Options`

Reply:
I think we need to sort out the neutrality of these bits too, since options currently co-mingles AWS specific and vendor agnostic parameters. Let's spend some time pair programming this.